### DISCUSSION TITLE
Support Biome as formatter

### DISCUSSION THREAD

**Discussion by jase88** [nrwl/nx#23347] [Feature Requests]
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description

[Biome](https://biomejs.dev/) (previously Rome) made some news by reaching [97% compatibility with Prettier](https://console.algora.io/challenges/prettier).
A lot of rules from major plugins are already supported.

## Motivation
It would be beneficial performance-wise.
And by making the formatter plugable it would make nx even more future-proof.

## Suggested Implementation
official biome support with:
- migration from prettier configuration to biome config
- generator to provide configuration per project
- executor for running nx format with biome formatter

## Alternate Implementations
- provide an API to make formatting via plugin possible

**Comment by vaibhavdugar** [comment]
How to use biome in nx monorepo?

**Comment by guillempuche** [comment]
@vaibhavdugar, I have a repo for this https://github.com/guillempuche/nx-expo-next-tamagui/blob/main/biome.json

I still need to run lint for all the files (as @JamesHenry pointed https://github.com/guillempuche/nx-expo-next-tamagui/issues/1), but when coding a file on VSCode, it works.

**Comment by vaibhavdugar** [comment]
@guillempuche in your repo as well biome is not configured with nx, I see that you are also using ESLint with nx, do you have any idea how can we use biome instead of ESLint in nx.json?

**Comment by guillempuche** [comment]
There aren't Eslint nor Prettier dependencies in the root package.json.

It's all Biome.

I don't use Nx target lint (I should delete this). Instead, you can format the code via the script in package.json. There's also a pre-commit Husky script.

**Comment by vaibhavdugar** [comment]
Thanks, can you tell me how to run the biome command for linting to catch all the errors?

**Comment by guillempuche** [comment]
Two ways:

a) `yarn lint` and it alerts you for errors
b) commit a change, auto lint and it alerts you for errors. How? Husky runs `yarn pre-commit` that triggers dependency `lint-staged` https://github.com/guillempuche/nx-expo-next-tamagui/blob/main/.lintstagedrc

**Comment by MrChrisRodriguez** [comment]
We just updated our lint target as follows:

```
"lint": {
      "executor": "nx:run-commands",
      "options": {
          "command": "yarn dlx @biomejs/biome check --apply {projectRoot}"
      }
    },
```

**Comment by shinebayar-g** [comment]
We use https://nx.dev/features/enforce-module-boundaries. I guess we'll wait until nx officially supports biome.

**Reply by AlonMiz** [reply]
actually, that's our main blocker for us to move to biome

**Comment by jase88** [comment]
Biome has also a issue about monorepos: https://github.com/biomejs/biome/issues/2228 aiming to support a root configuration that is extendable per project.

**Comment by mcfdez** [comment]
Hello friends,

I am working on this project https://github.com/GitOpsLovers/nx-biome to create a plugin to implement Biome as linter, executor and generator in NX. Who wants to participate is welcome!

**Comment by guillempuche** [comment]
@mcfdez , how does it support child `biome.json` files https://biomejs.dev/guides/big-projects/?

**Comment by mcfdez** [comment]
> @mcfdez , how does it support child `biome.json` files https://biomejs.dev/guides/big-projects/?

Hi @guillempuche 

The idea is that the plugin configuration generator will create a biome.json in the project that you tell it and also creates a biome.json in the workspace root. In the biome.js of the project, an  _extend_  will be made.

**Comment by FrozenPandaz** [comment]
I've converted this to a discussion so it's easier to discuss.

**Comment by terrymun** [comment]
For performance reasons we are also thinking of migrating away from eslint + prettier and use biome instead. The only thing that concerns us is that there is no biome ruleset equivalent for the `@nx/eslint-plugin`.

For the time being, a possible solution is to have eslint enabled specifically for this plugin, so that we can still stop users from accidental cross module imports + run everything else with biome. Does anybody have any luck with this so far? And if this practice is recommended?

**Reply by FranciscoKloganB** [reply]
@terrymun I use `@nx/eslint` (without prettier) and biome together, I would say 99% successfully. The 1% that's missing it's probably my fault, I think I have two commands in `package.json` that are slightly differently defined, so one of them does not catch some of the files I wanted... :) 

Anyway, add this as the last override entry in your root `.eslintrc.json` and install `eslint-plugin-biome`.

```json
{
   "overrides": [
      {
         "extends": ["biome"],
         "files": ["*.cjs", "*.cts", "*.mjs", "*.mts", "*.js", "*.jsx", "*.ts", "*.tsx"],
         "rules": {}
      }
   ]
}
 ```
 
 Then add a `biome.json` to the root. The rest is pretty much the same old, same old. As in, if you have some specific rules in a given `app` or `lib`/`package` you can do add a biome.json inside it and do:
 
 ```json
{
  "$schema": "../../../node_modules/@biomejs/biome/configuration_schema.json",
  "extends": ["../../../biome.json"],
  "overrides": []
}
```

**Note:** I have intentionally kept `biome` outside `nx` control flow until one of these two happens (ideally both): `biome` matures for monorepos, `nx` officially adopts `biome`. However, so far I had no issues with my setup. Biome is quite fast and they have their own `affected` equivalent flag. So... All good.

**Reply by guillempuche** [reply]
For me, Nx + Biome - Eslint - Prettier is working great. Just one Biome config file in the root Nx directory until Biome supports monorepo properly.

My code base isn't big enough right now for packages Biome config files. Else, and divergent linting between packages, you could need Eslint.

**Comment by dsod** [comment]
I haven't had the need for project specific configuration files with Biome yet, but running it during CI and pre-commit hooks works great:

**CI**
`biome ci --changed --no-errors-on-unmatched --since=$NX_BASE`

Where `$NX_BASE` is set by the `nrwl/nx-set-shas@v4` action.

**Pre-commit hook script**
```bash
#!/bin/sh
set -eu

if git status --short | grep --quiet '^MM'; then
  printf '%s\n' "ERROR: Some staged files have unstaged changes" >&2
  exit 1;
fi

git diff -z --staged --name-only --diff-filter=ACMR | \
    xargs -0 pnpm biome check --apply --no-errors-on-unmatched

git update-index --again
```

One outstanding issue/pain-point I've had when using NX + Biome is that many generators will install ESLint all over again (without an option to disable it). Slightly annoying having to remove files and ESLint related packages from `package.json` after running generators like this. Example of such generator: https://nx.dev/nx-api/node/generators/application#linter

**Comment by nodegin** [comment]
Is there any official plan to support?
It seem the issue is moved to discussion but no official member involved at all

**Comment by nhuray** [comment]
@vsavkin and the Nx team, is there a plan to support Biome as an optional replacement for Prettier and Eslint in the near future ?

The biome team is actively working now on supporting [Monorepo](https://github.com/biomejs/biome/issues/2228), so it might be the right time to federate efforts !

**Reply by pawel-twardziak** [reply]
I am for it too!

I could work on it, with pleasure :) And make Biome available in NX.

**Reply by nhuray** [reply]
@pawel-twardziak Any news you can share on that topic?

**Comment by loadingalias** [comment]
This feels stale. I've come over from the "feature request" to this discussion, which has been going on for over a year... is it being worked on or considered or any kind of official news?

**Comment by qcho** [comment]
Since Biome v2 monorepo is now supported.

https://biomejs.dev/guides/big-projects/#monorepo

Wonder if there is an opportunity to have an official `@nx/biome-plugin` now

**Reply by Nightbr** [reply]
Also interested for an official `@nx/biome-plugin` :eyes:

**Reply by luckyluggi** [reply]
me 2

**Comment by Forceres** [comment]
Any updates?

**Reply by seyfert** [reply]
See this blog post from a couple weeks ago:
https://nx.dev/blog/integrate-biome-in-20-minutes

**Reply by Forceres** [reply]
Yeah, I checked out this blog. Implemented that plugin + generator of biome.json and that works great, thanks.

**Reply by abejfehr** [reply]
this doesn't really solve that `nx format:write` is slow with a very large set of files

**Reply by nhuray** [reply]
There's now a community [plugin](https://github.com/berenddeboer/nx-plugins/tree/main/packages/nx-biome) referenced  in the [Plugin Registry](https://nx.dev/docs/plugin-registry#_top)
